### PR TITLE
5352: Additional Service GET

### PR DESCRIPTION
**Pull request by RuthBerry96**
Story: [AB#5352](https://buyingcatalog.visualstudio.com/c5f97979-5b03-4d10-ba8d-871d0526b408/_workitems/edit/5352)
Task: [AB#8158](https://buyingcatalog.visualstudio.com/c5f97979-5b03-4d10-ba8d-871d0526b408/_workitems/edit/8158)

### CODE REVIEW

File: src/NHSD.BuyingCatalogue.Solutions.Application/Persistence/AdditionalServiceReader.cs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+namespace NHSD.BuyingCatalogue.Solutions.Application.Persistence
+{
+    internal sealed class AdditionalServiceReader
+    {
+        private readonly IAdditionalServiceRepository _additionalServiceRepository;
+
+        public AdditionalServiceReader(IAdditionalServiceRepository additionalServiceRepository)
+        {
+            _additionalServiceRepository = additionalServiceRepository;
```
Comment (CallumBradburyBJSS) [inline src/NHSD.BuyingCatalogue.Solutions.Application/Persistence/AdditionalServiceReader.cs:16]: Throw if null? Or did we stop doing that in BAPI?

File: tests/NHSD.BuyingCatalogue.Solutions.Persistence.DatabaseTests/AdditionalServiceRepositoryTests.cs [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+            const string solutionId = "Sln1";
+
+            await CreateSolution(solutionId);
+            await CreateAdditionalService("Cat", solutionId);
+
+            var request =
+                await _additionalServiceRepository.GetAdditionalServiceBySolutionIdsAsync(
+                    new List<string> { solutionId }, CancellationToken.None);
+
+            request.Count().Should().Be(1);
```
Comment (CallumBradburyBJSS) [inline tests/NHSD.BuyingCatalogue.Solutions.Persistence.DatabaseTests/AdditionalServiceRepositoryTests.cs:56]: We should probably verify the content of the list as well as the count

File: tests/NHSD.BuyingCatalogue.Solutions.Persistence.DatabaseTests/AdditionalServiceRepositoryTests.cs [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+            await CreateSolution(solutionId1);
+            await CreateSolution(solutionId2);
+            await CreateAdditionalService("Cat1", solutionId1);
+            await CreateAdditionalService("Cat2", solutionId2);
+
+            var request =
+                await _additionalServiceRepository.GetAdditionalServiceBySolutionIdsAsync(
+                    new List<string> { solutionId1, solutionId2 }, CancellationToken.None);
+
+            request.Count().Should().Be(2);
```
Comment (CallumBradburyBJSS) [inline tests/NHSD.BuyingCatalogue.Solutions.Persistence.DatabaseTests/AdditionalServiceRepositoryTests.cs:74]: Same dealio here, let's check that it's returning what we expect

File: src/NHSD.BuyingCatalogue.Solutions.API/Controllers/AdditionalServiceController.cs [hunk 4]
```diff
@@ -10,39 +15,44 @@ namespace NHSD.BuyingCatalogue.Solutions.API.Controllers
     [Produces(MediaTypeNames.Application.Json)]
     public sealed class AdditionalServiceController : ControllerBase
     {
+        private readonly IMediator _mediator;
+
+        public AdditionalServiceController(IMediator mediator)
+        {
+            _mediator = mediator ?? throw new ArgumentNullException(nameof(mediator)); ;
```
Comment (CallumBradburyBJSS) [inline src/NHSD.BuyingCatalogue.Solutions.API/Controllers/AdditionalServiceController.cs:22]: double semicolon at the end of this line

File: src/NHSD.BuyingCatalogue.Solutions.API/Controllers/AdditionalServiceController.cs [hunk 5]
```diff
@@ -10,39 +15,44 @@ namespace NHSD.BuyingCatalogue.Solutions.API.Controllers
...
-            return additionalServiceResults;
-        }
+            var additionalServices = (await _mediator.Send(new GetAdditionalServiceBySolutionIdsQuery(solutionIds))).ToList();
 
-        private static string SetAdditionalServiceId(string solutionId)
-        {
-            var additionalServiceId = string.Empty;
-            additionalServiceId += $"{solutionId}A001";
+            if (additionalServices.Count is 0)
+                return NotFound();
```
Comment (CallumBradburyBJSS) [inline src/NHSD.BuyingCatalogue.Solutions.API/Controllers/AdditionalServiceController.cs:42]: I'm not sure that we should return NotFound here, an empty list is probably better. NotFound implies the solution ids weren't real, but they were... they just didn't have any additional services. Hmm. It's a tricky one. Might be worth checking with @sam-roberts-bjss

File: src/NHSD.BuyingCatalogue.Solutions.Application/Persistence/AdditionalServiceReader.cs (hunk 1 above)
Comment (RuthBerry96) [reply src/NHSD.BuyingCatalogue.Solutions.Application/Persistence/AdditionalServiceReader.cs:16]: I cant seem to find any examples where we have thrown in the reader, but I can add it in if you want

File: tests/NHSD.BuyingCatalogue.Solutions.Persistence.DatabaseTests/AdditionalServiceRepositoryTests.cs (hunk 2 above)
Comment (RuthBerry96) [reply tests/NHSD.BuyingCatalogue.Solutions.Persistence.DatabaseTests/AdditionalServiceRepositoryTests.cs:56]: 👍

File: src/NHSD.BuyingCatalogue.Solutions.API/Controllers/AdditionalServiceController.cs (hunk 4 above)
Comment (RuthBerry96) [reply src/NHSD.BuyingCatalogue.Solutions.API/Controllers/AdditionalServiceController.cs:22]: 👍

File: src/NHSD.BuyingCatalogue.Solutions.API/Controllers/AdditionalServiceController.cs (hunk 5 above)
Comment (sam-roberts-bjss) [reply src/NHSD.BuyingCatalogue.Solutions.API/Controllers/AdditionalServiceController.cs:42]: Yeah Ruth & I talked about this & agreed to return an empty list. Also I think it should return all additional services if no solution ID is supplied. Agreed to raise this as tech debt, there isn't a requirement for this yet.

File: tests/NHSD.BuyingCatalogue.Solutions.Persistence.DatabaseTests/AdditionalServiceRepositoryTests.cs (hunk 3 above)
Comment (RuthBerry96) [reply tests/NHSD.BuyingCatalogue.Solutions.Persistence.DatabaseTests/AdditionalServiceRepositoryTests.cs:74]: 👍

File: src/NHSD.BuyingCatalogue.Solutions.Application/Persistence/AdditionalServiceReader.cs (hunk 1 above)
Comment (CallumBradburyBJSS) [reply src/NHSD.BuyingCatalogue.Solutions.Application/Persistence/AdditionalServiceReader.cs:16]: if we don't do it in BAPI, it's all good

**Pull request merged**